clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackSecurityGroupCachingAgentSpec.groovy

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2016 Target, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.clouddriver.openstack.provider.agent
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.cats.agent.CacheResult
+import com.netflix.spinnaker.cats.provider.ProviderCache
+import com.netflix.spinnaker.clouddriver.model.AddressableRange
+import com.netflix.spinnaker.clouddriver.model.securitygroups.IpRangeRule
+import com.netflix.spinnaker.clouddriver.model.securitygroups.Rule
+import com.netflix.spinnaker.clouddriver.openstack.cache.Keys
+import com.netflix.spinnaker.clouddriver.openstack.client.OpenstackClientProvider
+import com.netflix.spinnaker.clouddriver.openstack.model.OpenstackSecurityGroup
+import com.netflix.spinnaker.clouddriver.openstack.provider.OpenstackInfrastructureProvider
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackCredentials
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackNamedAccountCredentials
+import org.openstack4j.model.compute.IPProtocol
+import org.openstack4j.openstack.compute.domain.NovaSecGroupExtension
+import spock.lang.Specification
+import spock.lang.Subject
+
+import static com.netflix.spinnaker.clouddriver.openstack.cache.Keys.Namespace.SECURITY_GROUPS
+
+class OpenstackSecurityGroupCachingAgentSpec extends Specification {
+
+  @Subject
+  OpenstackSecurityGroupCachingAgent cachingAgent
+  OpenstackNamedAccountCredentials namedAccountCredentials
+  OpenstackCredentials credentials
+  OpenstackClientProvider provider
+  ObjectMapper objectMapper
+  ProviderCache providerCache
+  String region = 'east'
+  String accountName = 'os-account'
+
+  def setup() {
+    providerCache = Mock(ProviderCache)
+    provider = Mock(OpenstackClientProvider)
+    credentials = GroovyMock(OpenstackCredentials)
+    credentials.provider >> provider
+    namedAccountCredentials = Mock(OpenstackNamedAccountCredentials)
+    namedAccountCredentials.credentials >> credentials
+    objectMapper = Mock(ObjectMapper)
+    cachingAgent = Spy(OpenstackSecurityGroupCachingAgent, constructorArgs: [namedAccountCredentials, region, objectMapper])
+    _ * cachingAgent.getAccountName() >> accountName
+  }
+
```

Comment:
What do you want tested for exceptions? `OpenstackProviderException` could be thrown, but it will just bubble up.